### PR TITLE
`Development`: Add a shared build job queue for local continuous integration

### PR DESCRIPTION
**Pull request by laurenzfb**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://ls1intum.github.io/Artemis/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

The Local CI build job queue is not distributed. For build jobs to be distributed accross server instances and being able to display the current build job queue in the client, we need a queue that is shared accross the server instances.

### Description

This PR implements a Hazelcast IQueue that can be seen and accessed by Hazelcast cluster members. Whenever a build job is triggered, a build job item including the participation id and if existing, the commit hash, is added to the queue. The service actively listens for new items in the queue. If a new item is added, the service checks whether the server instance has the capacity to process the build job. If so, the build job is processed. 
I adjusted some test cases to account for changes.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- Local VC CI Test Server (3 or 6)

1. As Instructor, create a programming exercise
2. Check, whether build jobs are processed correctly by confirming the build results
3. As Student, make a submission
4. Confirm build result

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### CODE REVIEW

**Review by basak-akan — DISMISSED** [review]
Tested on TS3, build results are confirmed for both instructor and student

**Review by max-bergmann — DISMISSED** [review]
Tested on TS3. LGTM

**Review by milljoniaer — DISMISSED** [review]
Tested on ts3, works as expected

**Review by vinceclifford — DISMISSED** [review]
Tested on TS3. Works as intended

**Review by milljoniaer — APPROVED** [review]

**Review by vinceclifford — APPROVED** [review]
Reapproved

**Review by max-bergmann — APPROVED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCISharedBuildJobQueueService.java [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        if (buildJob != null) {
+            Long participationId = buildJob.getParticipationId();
+            buildJob.setExpirationTime(System.currentTimeMillis() + TimeUnit.SECONDS.toMillis(130));
+            processingJobs.put(participationId, buildJob);
+        }
+        return buildJob;
+    }
+
+    private ProgrammingExerciseParticipation retrieveParticipationWithRetry(Long participationId) {
+        int maxRetries = 5;
```
Comment (reschandreas) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCISharedBuildJobQueueService.java:183]: you could make this final, but that's non-blocking for me

**Review by reschandreas — APPROVED** [review]
code lgtm 👍🏻

**Review by basak-akan — APPROVED** [review]
code looks good

**Pull request merged**